lib/internal/util.js

```diff
@@ -187,3 +187,18 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+exports.getConstructorOf = function getConstructorOf(obj) {
+  while (obj) {
+    var descriptor = Object.getOwnPropertyDescriptor(obj, 'constructor');
```

Comment:
The performance of `Object.getOwnPropertyDescriptor` is about 17 times slower than `Object.prototype.hasOwnProperty.call`. Perhaps the code from getEligibleConstructor can be used instead.